### ISSUE TITLE
TypeError: Ajv is not a constructor with Jest 27

### ISSUE THREAD

**Issue by nickmccurdy** [eslint/eslint#15030]
Related to https://github.com/eslint/eslintrc/issues/21

**Tell us about your environment**

* **ESLint Version:** 7.28.0
* **Node Version:** 16.8.0
* **npm Version:** 7.22.0
* **Operating System:** macOs 11.5.2

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?**

`@babel/eslint-parser`

**Please show your full configuration:**

https://github.com/testing-library/eslint-plugin-jest-dom/blob/5cfb8c97ecc78fb4f61748fe083123bed7dc1f9e/package.json#L58-L66

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
import { RuleTester } from "eslint";
import * as rule from "../../../rules/rule";

const ruleTester = new RuleTester({ parserOptions: { ecmaVersion: 2015 } });
ruleTester.run("rule", rule, { ... });
```

```bash
npm run validate
```

**What did you expect to happen?**
No error (tests pass on Jest 26)

**What actually happened? Please copy-paste the actual, raw output from ESLint.**
`TypeError: Ajv is not a constructor`

**Steps to reproduce this issue:**

```
git clone https://github.com/testing-library/eslint-plugin-jest-dom
cd eslint-plugin-jest-dom
git checkout 5cfb8c97ec
npm install
npm run validate
```

**Are you willing to submit a pull request to fix this bug?**
It seems like the issue is caused by ESLint requiring two different versions of `ajv` in the tree, but I don't know how to fix that.

**Comment by mdjermanovic** [comment]
Hi @nickmccurdy, thanks for the issue!

This seems to be a bug in jest's resolver.

I first thought that it's mistakenly resolving `require("ajv")` in `node_modules/@eslint/eslintrc/lib/shared/ajv.js` to Ajv 8 (which is a dependency of the `table` package, and does have a different API) instead of Ajv 6, so I did some debugging and added the following after `const Ajv = require("ajv")` in `node_modules/@eslint/eslintrc/lib/shared/ajv.js` :

```js
if (typeof Ajv !== "function") {
    throw new Error(`
        type: ${typeof Ajv}
        own: { ${Object.getOwnPropertyNames(Ajv)} }
        __proto__: { ${Object.getOwnPropertyNames(Object.getPrototypeOf(Ajv))} }
        __proto__.__proto__: ${Object.getPrototypeOf(Object.getPrototypeOf(Ajv))}
        require.resolve("ajv"): ${require.resolve("ajv")}
    `);
}  
```

Then I ran `npm run validate`. The linting step goes well, but testing fails with:

```
[test]             type: object
[test]             own: {  }
[test]             __proto__: { constructor,__defineGetter__,__defineSetter__,hasOwnProperty,__lookupGetter__,__lookupSetter__,isPrototypeOf,propertyIsEnumerable,toString,valueOf,__proto__,toLocaleString }
[test]             __proto__.__proto__: null
[test]             require.resolve("ajv"): .../eslint-plugin-jest-dom/node_modules/@eslint/eslintrc/lib/shared/ajv.js
```

So it looks like it's resolving `require("ajv")` to the module itself (which happens to be named 'ajv.js') as if it was `require("./ajv")`, and returns an empty object.

**Comment by nickmccurdy** [comment]
Thanks for the info. I'm having trouble finding a straightforward reproduction for the issue with only Jest, but if I use `kcd-scripts` (specifically the `kcd-scripts test` command which has its own Jest config) I can reproduce the issue.

**Comment by nickmccurdy** [comment]
I've narrowed down the issue to our Jest config, which adds `'shared'` to `moduleDirectories`, causing the import of `ajv` to look for the same file in `shared` rather than the package in `node_modules`.

This seems to be unrelated to ESLint as a default Jest config will work fine, so I'll close the issue. Sorry for the noise.
